[perl] Add annotator README.md in perl/ (#581)
bystrogenomics/bystro: perl/ANNOTATION.md

# Annotation Fields (Human Assembly hg38 and hg19)

<h2>Annotation Field Description</h2>

> ##### _Italicized fields_ are custom Bystro fields. All others are sourced as described.

Reviewer: None of the fields below seem to be italicized

<br/>

#### Summary of the Bystro Annotator

> Bystro takes 1+ VCF or SNP (PEMapper/PECaller) files, performs quality control, generates statistics, and outputs an annotated file that is in an easy-to-parse tab-separated format.

Reviewer: 1+ is a little bit of a confusing way to read this. I might just write out 'one or more'

> Bystro takes 1+ VCF or SNP (PEMapper/PECaller) files, performs quality control, generates statistics, and outputs an annotated file that is in an easy-to-parse tab-separated format.

Reviewer: ```suggestion
Bystro takes one or more VCF or SNP (PEMapper/PECaller format) files, performs quality control, generates statistics, and outputs an annotated file that is in an easy-to-parse tab-separated format.
```

> Each row in the output file corresponds to a single variant site, and contains a set of annotations that describe the site's genomic context, functional impact, allele frequencies in various populations, and more.

Reviewer: Lmk if this is incorrect, but doesn't it refer to a single variant? I.e. if you have multiple variants at a single site, they would be on different lines, right? If you had a multi-allelic or indels at the exact same variant site, they would be on multiple lines? If this is the case, would just drop 'site' for clarity.

This structure enables precise retention of complex data relationships across multiple annotation fields while maintaining a highly structured and parseable output.

#### What Information Can Bystro Output?

Bystro is a general-purpose annotation engine, and has no restrictions on the types of annotations/feature labels it can output. The annotations are sourced from the Bystro Database chosen, and is described based on a YAML configuration file, which describes the fields to be output, and the sources of the annotations.

A Bystro Database is a high-performance key-value database that uses the Lightning Memory Map Database (LMDB) engine. It supports millions of lookups per second on a single machine, and can be used to store and retrieve annotations for millions of variants.

Bystro Databases can be re-created from the YAML configuration file corresponding to that database, and new databases with different information can be created by editing the YAML configuration file, and re-running the Bystro Database creation process.

#### Variant Representation

> Bystro's variant representation deviates slightly from the standard VCF format. Namely, Bystro ensures that the reference base is always a single nuleotide, and does the appropriate normalization/transformation of input variants to ensure this. This variant representation greatly simplifies downstream analysis.

Reviewer: ```suggestion
Bystro's variant representation deviates slightly from the standard VCF format. Namely, Bystro ensures that the reference base is always a single nucleotide, and does the appropriate normalization/transformation of input variants to ensure this. This variant representation greatly simplifies downstream analysis.
```

Compared with the VCF format, Bystro's variant representation has the following differences:

SNPs: Unchanged

> Insertions: The reference base is always the base just before the insertion, and the alternate allele is the inserted sequence, preceeded by a +. The position is the first affected base.

Reviewer: ```suggestion
Insertions: The reference base is always the base just before the insertion, and the alternate allele is the inserted sequence, preceded by a +. The position is the first affected base.
```

> - For instance, if the VCF had POS: 1 REF: AT and ALT: ACCT, Bystro will represent this as inputRef: T, alt: +CCT, pos: 1, with the position the same as that in the VCF file, which is the first position of the reference base (the base just before the first inserted base). In Bystro's case there is only 1 reference base always, so the position is that of the only affected base.

Reviewer: For visual clarity, I think I would break the examples out into code block lines throughout this explanation, and same for the one below.

Deletions: VCFs require ALT alleles to be at least 1 nucleotide long, which means that to represent a deletion, the VCF has to "pad" the reference. Bystro removes this padding, and represents deletions as the reference base, followed by a - and the number of deleted bases. The position is the first affected base.

- For instance if the VCF had POS:1 REF: AT and ALT: A, Bystro will represent this as inputRef: T, alt: -1, pos: 2, notably shifting the position of the allele up by 1 base, since the deletion did not actually occur at position 1, but the next base over.

> Multiallelics: Bystro decomposes multiallelic sites into separate rows, each with a single alternate allele. This ensures that each row has a single alternate allele, and that the annotations are accurate for each allele. The appropriate transformations are done to ensure that VCF padding rules are correctly transformed into Bystro's representation, and ensures that all variants outtputted in Bystro are left-aligned.

Reviewer: I would recommend a quick example of this too, to promote quicker understanding

`alt` - the alternate/nonreference allele

- VCF multi-allelic and MNP sites are decomposed into individual entries of a single allele.
  - Genotypes are properly segregated per allele

`trTv` - transition:transversion ratio for your dataset at this position

- Possible values: `0`, `1`, `2`
  - 0 indicates neither transition nor transversion (which occurs when the alternate allele is an insertion or deleetion)
  - 1 is a transition (purine -> purine or pyrimidine -> pyrimidine)
  - 2 is a transversion (pyridine -> pyrimidine or vice versa)

`heterozygotes` - The array of heterozygous sample labels

`heterozygosity` - The fraction of all samples (excluding missing samples) that are heterozygous for the alternate allele

`homozygotes` - The array of homozygous sample labels

`homozygosity` - The fraction of all samples that are homozygous for the alternate allele

`missingGenos` - The samples that did not have a genotype (e.g., ".") at the site. If an individual has at least 1 missing genotype, they are considered missing for the site.

> - e.g., .|., .|0 are all considered missing

Reviewer: I would drop the e.g., here unless you mean ',' is considered missing. I like the 'Possible values:' at line 91, I think that format is just a little clearer to grasp right away. Might be nice to have 'possible values' wherever it's just a few options like that, or 'example values' when there's more possibilities and it's easy to show an example

`missingness` - The fraction of all samples that have missing genotypes for the alternate allele

`ac` - The alternate allele count

> `an` - The total non-mising allele count

Reviewer: ```suggestion
`an` - The total non-missing allele count
```

`sampleMaf` - The in-sample alternate allele frequency

`vcfPos` - The original VCF `POS`, unaffected by Bystro normalization transformations

`id` - The VCF `ID` field, if any

`discordant` - TRUE if the reference base provided in the input VCF matches the Bystro-annotated UCSC reference, FALSE otherwise

`ref` - The Bystro-annotated reference base(s), from the 'ref' track in the Bystro Database

- In the default Bystro Database, this is sourced from the UCSC reference genome
- In custom Bystro Databases, this can be sourced from any reference genome
- In the case of insertions the `ref` will be 2 bases long, the base just before the insertion, and the one right after
- In the case of deletions, the ref will be as long as the deletion, up to 32 bases (after that, the ref will be truncated)

<br/>

#### Transcript Annotations

In the default Bystro Database, we source transcript annotations from the UCSC refGene track, joined on other UCSC tracks: knownToEnsembl, kgXref, knownCanonical.

> - See [refGene]('https://sc-bro.nhlbi.nih.gov/cgi-bin/hgTables?hgsid=554_JXUlabut7OUQtCyNphC8FGaeUJnj&hgta_doSchemaDb=hg38&hgta_doSchemaTable=refGene') and [kgXref]('https://sc-bro.nhlbi.nih.gov/cgi-bin/hgTables?hgsid=554_JXUlabut7OUQtCyNphC8FGaeUJnj&hgta_doSchemaDb=hg38& hgta_doSchemaTable=kgXref') for more information

Reviewer: kgXref link is broken, I think it's just that there's a space in there

Reviewer: refGene link also doesn't seem to work

`refSeq.codonPosition` - The site's position within the codon (1, 2, 3)

`refSeq.codonNumber` - The codon number within the transcript

`refSeq.strand` - The positive or negative (Watson or Crick) strand

`refSeq.name` - RefSeq transcript ID

`refSeq.name2` - RefSeq gene sybmol

`refSeq.description` - The long form description of the RefSeq transcript

`refSeq.kgID` - UCSC's <a href='https://www.ncbi.nlm.nih.gov/pubmed/16500937' target='_blank'>Known Genes</a> ID

`refSeq.mRNA` - The mRNA ID, the transcript ID starting with NM\_

`refSeq.spID` - <a href="http://www.uniprot.org" target='_blank'>UniProt</a> protein accession number

`refSeq.spDisplayID` - <a href="http://www.uniprot.org" target='_blank'>UniProt</a> display ID

`refSeq.protAcc` - NCBI protein accession number

`refSeq.rfamAcc` - <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC165453/" target="_blank">Rfam</a> accession number

`refSeq.tRnaName` - Name from the tRNA track

`refSeq.ensemblID` - The Ensembl transcript id

`refSeq.isCanonical` - Whether the transcript is the canonical splice variant for the gene

<br/>

#### nearest.refSeq

> The nearest transcript(s), calculated by trascript start, transcript end boundaries. Transcripts that are equidistant are all outputted.

Reviewer: ```suggestion
The nearest transcript(s), calculated by transcript start, transcript end boundaries. Transcripts that are equidistant are all outputted.
```

`clinvarVcf.AF_TGP`: Allele frequencies from TGP

`clinvarVcf.ALLELEID`: The ClinVar Allele ID

`clinvarVcf.CLNDN`: ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB

`clinvarVcf.CLNDNINCL`: For included Variant : ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB

`clinvarVcf.CLNHGVS`: Top-level (primary assembly, alt, or patch) HGVS expression

`clinvarVcf.CLNREVSTAT`: ClinVar review status of germline classification for the Variation ID

`clinvarVcf.CLNSIG`: Aggregate germline classification for this single variant;

`clinvarVcf.CLNSIGCONF`: Conflicting germline classification for this single variant

`clinvarVcf.CLNVCSO`: Sequence Ontology id for variant type

`clinvarVcf.DBVARID`: NSV accessions from dbVar for the variant

`clinvarVcf.ORIGIN`: Allele origin. One or more of the following values may be added: 0 - unknown; 1 - germline; 2 - somatic; 4 - inherited; 8 - paternal; 16 - maternal; 32 - de-novo; 64 - biparental; 128 - uniparental; 256 - not-tested; 512 - tested-inconclusive; 1073741824 - other

`clinvarVcf.RS`: dbSNP ID (i.e. rs number)

> // add logofunc.id logofunc.alt logofunc.prediction logofunc.neutral logofunc.gof logofunc.lof genebass.id genebass.alt genebass.phenocode genebass.description

Reviewer: Looks like you added these below so I think you can drop this line

<br/>

#### (hg38-only) [LoGoFunc](

> https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10688473/)

Reviewer: Just checking, should this be on the same line as the title for this section or is there a reason why it's below?

A machine learning method for predicting pathogenic GOF, pathogenic LOF, and neutral genetic variants, trained on a broad range of gene-, protein-, and variant-level features describing diverse biological characteristics.

`logofunc.id`: The LoGoFunc VCF `ID`

`logofunc.alt`: The Bystro VCF-preprocessor's ALT record for this LoGoFunc site. This should always match the row's `alt` field value

`logofunc.prediction`: The LoGoFunc prediction

`logofunc.neutral`: The LoGoFunc neutral score

`logofunc.gof`: The LoGoFunc gain of function (GOF) score

`logofunc.lof`: The LoGoFunc loss of function (LOF) score

#### (hg38-only) [GeneBass](<https://www.cell.com/cell-genomics/pdf/S2666-979X(22)00110-0.pdf>)

> 

Reviewer: I would add a small description for GeneBass

`genebass.id`: The GeneBass VCF `ID`

`genebass.alt`: The Bystro VCF-preprocessor's ALT record for this GeneBass site. This should always match the row's `alt` field value

`genebass.phenocode`: The GeneBass phenotype code

> `genebass.description: The GeneBass description

Reviewer: ```suggestion
`genebass.description`: The GeneBass description
```
